crates/store/src/accounts/mod.rs

```diff
@@ -0,0 +1,490 @@
+
+        let leaf_index = NodeIndex::from(leaf.index());
+
+        // Apply reversion overlays to reconstruct historical state
+        let (path, leaf) = Self::apply_reversion_overlays(
+            &self.overlays,
+            block_target,
+            path_nodes,
+            leaf_index,
+            leaf,
+        )?;
+
+        // Extract commitment from leaf
+        let commitment = match leaf {
+            SmtLeaf::Empty(_) => EMPTY_WORD,
+            SmtLeaf::Single((_, value)) => value,
+            SmtLeaf::Multiple(_) => unreachable!("AccountTree uses prefix-free IDs"),
+        };
+
+        AccountWitness::new(account_id, commitment, path).ok()
+    }
+
+    /// Initializes the path nodes array from the latest state.
+    ///
+    /// Nodes that match their depth's empty subtree root are stored as `EMPTY_WORD` sentinel.
+    /// For non-empty depths, we store the actual node value.
+    fn initialize_path_nodes(path: &SparseMerklePath) -> [Word; SMT_DEPTH as usize] {
+        let mut path_nodes = [EMPTY_WORD; SMT_DEPTH as usize];
+
+        let path = Vec::from_iter(path.iter());
+
+        path.iter().rev().enumerate().for_each(|(depth, &node)| {
+            let empty_root = *EmptySubtreeRoots::entry(SMT_DEPTH, (depth + 1) as u8); // +1: depth from leaf
+            if node != empty_root {
+                path_nodes[depth] = node;
+            }
+        });
+
+        path_nodes
+    }
+
+    /// Applies reversion overlays to reconstruct the historical state.
+    ///
+    /// Iterates through overlays from newest to oldest (walking backwards in time),
+    /// updating both the path nodes and the leaf value based on reversion mutations.
+    /// Uses `EMPTY_WORD` as a sentinel to indicate "use empty subtree root for this depth".
+    fn apply_reversion_overlays(
+        overlays: &BTreeMap<BlockNumber, HistoricalOverlay>,
+        block_target: BlockNumber,
+        mut path_nodes: [Word; SMT_DEPTH as usize],
+        leaf_index: NodeIndex,
+        mut leaf: SmtLeaf,
+    ) -> Option<(SparseMerklePath, SmtLeaf)> {
+        // Iterate through overlays in reverse (newest to oldest)
+        for (_, overlay) in
+            overlays.iter().rev().take_while(|(block_num, _)| block_target <= **block_num)
+        {
+            // Update path sibling nodes that changed in this overlay
+            for sibling in leaf_index.proof_indices() {
+                let height = sibling
+                    .depth()
+                    .checked_sub(1) // -1: Convert from 1-indexed to 0-indexed
+                    .expect("proof_indices should not include root")
+                    as usize;
+
+                // Apply reversion mutation if this node was modified
+                // EMPTY_WORD sentinel means "use empty subtree root for this depth"
+                if let Some(hash) = overlay.node_mutations.get(&sibling) {
+                    path_nodes[height] = *hash;
+                }
```

Comment:
Related to one of the above comments: if in `HistoricalOverlay::new()` we use empty subtree roots instead of `EMPTY_WORD`, we should be able to avoid using `EMPTY_WORD` as a sentinel value, and would also not need the `build_dense_path()` helper.